[ripper37-libbase] Add new port

RippeR37 opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

For now it is intended that this is only statically-linked library. I may add support for building as a shared lib in the future version (like v1.1 or something like that).

RippeR37 committed:
Add libbase (v1.0.1) to vcpkg

RippeR37 committed:
version database

dg0yt commented:
Early community feedback: I guess the name "libbase" will not be accepted.
- It is very generic.
- It is already taken: https://repology.org/project/libbase/versions

You might simply name the port `ripper37-libbase`. But file name issues might still be a problem. Seeing the mention of chrome's //base: IIRC we already struggled with how it proliferates into various packages. And namespace `base` also isn't very distinct.

RippeR37 commented:
Hi @dg0yt and thank you for a quick feedback!

The library description is pretty much

> standalone (no/very limited dependencies) reimplementation of `//base` from Chromium

so it would be hard for me to come up with a better and more descriptive name for it that people familiar with Chromium codebase (main target audience) would understand/recognize (the `base` is one thing people might "get"). Similarly namespace or include path is selected specifically to reduce friction for users (I'm providing same/similar APIs with same/similar includes to original implementation) that deal with Chromium codebase normally so they can write "same" code within Chromium project and in their own separate projects.

I see there is one entry for `libbase` in repology although its hard for me to track what it is exactly, doesn't seem like an actual C++ package that could be added to vcpkg at future time and either related to something in java or URLs [lead](https://github.com/deepin-community/libbase) to [404 pages](http://reporting.pentaho.org/libbase). Unfortunately this isn't custom project from the scratch so for me choosing name for it was kinda forced by the nature of what the library provides.

If the port won't be allowed under that name, I guess I can use different port name (suggestions welcome what would be acceptable and still close to the original name so people looking for it will know it's that) but I don't think I would want to change the name of the library itself if that would be required.

I also wanted to add that I did some extra work to ensure that installed files (includes) aren't in `base` folder but `libbase/base` to ensure that there won't be _collisions_ with other C++ libraries (as `base` could a much more common include dir than `libbase`).

RippeR37 commented:
Updated to `36ca003ef5b1b8b06a7c55c9d26e8b2899c34557` with:

* added rule to remove `.../debug/share` files (caused failures in CI in post build checks)
  *  `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")`
* slight update to the usage file

Not sure what is the required structure of the commits in PRs (if they will be squashed together) so I updated them in-place and force-pushed new versions instead of adding fixup commits. Hope its good.

RippeR37 committed:
Empty commit (to re-run tests)

Cheney-W commented:
@RippeR37 
1. I think the name of this port needs to be changed to `ripper37-libbase`, at the very least, it shouldn't be `libbase`.
2. The line 32 of portfile.cmake should be changed as below:
```
file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```
  I know that the `CMakeLists.txt` of this port already contains code to install the LICENSE. Adding it in `portfile.cmake` is just to 
  prevent issues in case `CMakeLists.txt` changes.

3. libbase[unittests] install failed with below error:
```
CMake Error at F:/44554/downloads/tools/cmake-3.30.1-windows/cmake-3.30.1-windows-i386/share/cmake-3.30/Modules/GoogleTestAddTests.cmake:112 (message):
  Error running test executable.
    Path: 'F:/44554/buildtrees/libbase/x64-windows-dbg/tests/unit/libbase_unittests.exe'
    Working directory: 'F:/44554/buildtrees/libbase/x64-windows-dbg/tests/unit'
    Result: Exit code 0xc0000135
    Output:
Call Stack (most recent call first):
  F:/44554/downloads/tools/cmake-3.30.1-windows/cmake-3.30.1-windows-i386/share/cmake-3.30/Modules/GoogleTestAddTests.cmake:226 (gtest_discover_tests_impl)
```

RippeR37 commented:
Hey @Cheney-W 👋 

w.r.t. 1. & 2. - sure, I'll update it.

w.r.t 3. - could you help me a bit with it? The problem seems to be that when `libbase` is built by vcpkg with `unittests` feature enabled (basically also builds unittests) the unittests EXE doesn't get required DLLs copied into the dir with them. However when I build it locally myself, CMake does correctly copy the DLLs to EXE location and they run fine. I saw there is `vcpkg_add_to_path()` but it feels hacky (and not even sure if that would solve it).

My first guess was that exes/tools aren't supported, but it seems other packages do have tools/executables configured as features so not sure if I'm doing something wrong in the `libbase` project itself or if this is problem with port configuration.
I did some tests with installation config in CMake around unittest target but it doesn't seem to be that (or I'm still missing something). I could also just not expose unittests/perftests but I guess this problem could come up in the future with other optional "features" and there technically there is some merit if user wants to install those as well.

What would be the proper way to solve this within VCPKG?

Cheney-W commented:
Yes, this issue is caused by the presence of the `gtest_discover_tests(libbase_unittests)` statement in the port's `CMakeLists.txt` file.  

VCPKG supports the installation of executables/tools, but it generally does not run these newly built executables during the build process. Instead, they and their required DLLs are placed in the `tools` folder for users to use.  

To fix this issue, I think the following steps are necessary:  
1. Add a patch to prevent `libbase_unittests` from being executed during the build process.  
2. Modify the following three files to add installation rules for each target so that they are correctly placed in the appropriate VCPKG directories.  
```
libbase\src\v1.0.1-06ccb5e9e9\tests\unit\CMakeLists.txt
libbase\src\v1.0.1-06ccb5e9e9\tests\perf\CMakeLists.txt
libbase\src\v1.0.1-06ccb5e9e9\examples\simple\CMakeLists.txt
install(
   TARGETS libbase_unittests
   ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
   LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
   RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
)
```
4. Add the following content to `portfile.cmake` to store the newly generated executable files and their required DLLs by calling the `vcpkg_copy_tools` function.
```
if("examples" IN_LIST FEATURES)
    vcpkg_copy_tools(TOOL_NAMES simple AUTO_CLEAN)
endif()

if("unittests" IN_LIST FEATURES)
    vcpkg_copy_tools(TOOL_NAMES libbase_unittests AUTO_CLEAN)
endif()

if("perftests" IN_LIST FEATURES)
    vcpkg_copy_tools(TOOL_NAMES libbase_perf_tests AUTO_CLEAN)
endif()
```
Eventually, these executables will be placed in `tools\${PORT}`.
![image](https://github.com/user-attachments/assets/824456a1-b05e-4b0b-b105-bb98267bd0be)

dg0yt commented:
Wait...
- Building examples without installing them is pointless. But does the port really need this feature?
- Installing test executables is not desired, and it doesn't resolve the problem.
  You seem to be looking for the `ADD_BIN_TO_PATH` option:
  https://learn.microsoft.com/en-us/vcpkg/maintainers/functions/vcpkg_cmake_install#add_bin_to_path  
  Again, does the port really need this feature?

dg0yt commented on ports/libbase/usage:
```diff
@@ -0,0 +1,8 @@
+
+    find_package(libbase CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE libbase::libbase)
+
+For more information, see the documentation at:
+
+    https://ripper37.github.io/libbase/
```
Keep it brief. There is a manifest with that link.
```suggestion
```

dg0yt commented on ports/libbase/usage:
```diff
@@ -0,0 +1,8 @@
+The package libbase provides CMake targets:
+
+    find_package(libbase CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE libbase::libbase)
```
Standard pattern used by the heuristics from the tool.
```suggestion
libbase provides CMake targets:

  find_package(libbase CONFIG REQUIRED)
  target_link_libraries(main PRIVATE libbase::libbase)
```

RippeR37 committed:
Apply suggestions from code review

Co-authored-by: Kai Pastor <dg0yt@darc.de>

RippeR37 committed:
Renamed port 'libbase' to 'ripper37-libbase'

RippeR37 committed:
Updated vcpkg.json and portfile.cmake as per review

RippeR37 committed:
Updated versions files

RippeR37 commented:
> 2\. The line 32 of portfile.cmake should be changed as below:
> 
> 
> ```
> file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
> ```

BTW. if that is the recommended way to install usage file, consider updating the documentation ([this](https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-packaging?pivots=shell-powershell) and [that](https://learn.microsoft.com/en-us/vcpkg/maintainers/handling-usage-files)).

----

I also wasn't sure if I should change the `name` in `vcpkg.json` to `ripper37-libbase` or keep it `libbase`. For now, I did as I assumed it has to match port directory name, but I wonder if it won't interfere with anything else, so let me know if it should stay like that or if I should revert it to just `libbase`.

RippeR37 committed:
Updated versions database

Cheney-W commented on ports/ripper37-libbase/portfile.cmake:
```diff
@@ -0,0 +1,34 @@
+    HEAD_REF master
+)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+)
+
```
These codes should be removed.

RippeR37 replied:
Ack. I left it because my next version will contain actual optional feature within the library itself, but I can always bring it back later :) Removed now.

dg0yt commented:
> I based my version on the documentation ([this](https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-packaging?pivots=shell-powershell) and [that](https://learn.microsoft.com/en-us/vcpkg/maintainers/handling-usage-files)) so might be good to update it if so.

Thanks, I was already looking for where this comes from. https://github.com/microsoft/vcpkg-docs/pull/463

Cheney-W commented on ports/ripper37-libbase/usage:
```diff
@@ -0,0 +1,4 @@
+libbase provides CMake targets:
```
```suggestion
ripper37-libbase provides CMake targets:
```

RippeR37 committed:
Apply review fixes

RippeR37 committed:
Update versions database

Cheney-W commented on ports/ripper37-libbase/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+        -DLIBBASE_BUILD_EXAMPLES=OFF
+        -DLIBBASE_BUILD_TESTS=OFF
+        -DLIBBASE_BUILD_PERFORMANCE_TESTS=OFF
+        ${FEATURE_OPTIONS}
```
```suggestion
```
Sorry, I forgot to delete this variable.

RippeR37 replied:
Damn, my bad, sorry for missing that and thank you for noticing it.

RippeR37 committed:
Update portfile.cmake (remove not needed line)

RippeR37 committed:
Update versions database

dg0yt commented:
The non-Windows binary names come with a double prefix:
`ripper37-libbase:arm64-osx:/lib/liblibbase.a`

FTR I still think the name of binary (libbase.lib, libbase.a) has a significant potential for clashes. There are no immediate clashes, but any search for a `base libbase` namespec may give surprises. Even more on case-insensitive file systems. Context:

- `ripper37-libbase:arm64-windows:/lib/libbase.lib` (This PR.)
- `simd:x64-windows:/lib/Base.lib`
- `vst3sdk:x64-windows:/lib/vst3sdk/base.lib`
- `crashpad:x64-windows:/debug/lib/vcpkg_crashpad_base.lib` (renamed in the vcpgk port to resolve the clash)
- maybe other google projects, maybe skipped in CI.
- "base" libraries in user projects.

RippeR37 commented:
> The non-Windows binary names come with a double prefix:
> `ripper37-libbase:arm64-osx:/lib/liblibbase.a`

I'm fine with this and this should also reduce chances of clashes with other libraries quite a lot I'd assume (as its somewhat likely to have `base` lib, but unlikely to have `libbase` lib (`liblibbase.a`/`liblibbase.so`)).

dg0yt commented:
Opinion:

`base` and `libbase` are not really good choices for a standalone library name. Even less in vcpkg where CI installs everything at once (subject to order defined by dependencies), where msbuild autolinks everything in `lib` and where users and packages may have their own "base" libs.

The reason that there is (almost!) no other project using "base" for a standalone library is ... that they understood that it is not an idea which scales well? Or maybe they learned a lesson and moved away? There are app. 16 `<unique>base.lib` in x64-windows.

`liblib` isn't a sufficient disambiguation but just adds confusion. In general it indicates an intended build system error. You don't want to be an exeception to the rule.  (It is easy to rectify with target properties.)

But I'm not in the position to ask for changes.

RippeR37 commented:
Thank you. I get that and that makes perfect sense. But from my PoV this library is meant to be low level building block(s) next to the C++ standard library that you take when you start a project and build something on top of it, sort of like mini-framework. It's not meant to be used like supplementary library where it does one specific thing and you only get it for that task and it makes sense to be able to drop-in everywhere (e.g. json parser).

I also don't want to change namespace (function names, etc.) precisely because people familiar with Chromium's `//base` would have it hard time using it when they constantly have to fight against their "muscle memory" - I assume that would be very painful. The goal is to be able to create a new project, import this lib and start writing a small standalone project with all the nice things (one potentially knows from Chromium) with ease.

I understand that there could be clashes with different libs out there and I also think there will be/are already a lot of libs in vcpkg or overall in C++ community that have clashes with each other, especially in less obvious things like exporting functions with same names or includes with same folders. And I guess in the context of what this library is - it seems fine. One can always _not_ use `libbase` if it isn't compatible with other things they need or they can vendor/fork it switch quick grep or override via some defines if.

As for the artifact names - I don't care that much what are they named, especially in vcpkg port, as its not really exposed to users anyway. I think it might make a bit difference for debugging if the name would somehow come up in the stack traces/core dumps, but I'm not sure how confusing that would be to people (or if it would even be a problem?). So I'm mostly fine with overriding them in the vcpkg port, but I'd prefer to keep the usage of the library (i.e. CMake target names) same between vcpkg port and actual library because it would be more complicated for people to migrate from "vendoring" it to switching to vcpkg then. Is there an easy way to make vcpkg port override CMake to build `libripper37libbase.lib` (or libripper37base.lib`) without changing the CMake target {{libbase::libbase}}?

Also, if I understand correctly, there are some magic tricks to use custom vcpkg triples that allows e.g. overriding defines? If that's true, than if there would actually be people that have problem with this lib clashing with something else, I could expand it in the future to allow overriding namespaces (and possibly more) via that functionality.

dg0yt commented:
> One can always not use libbase if it isn't compatible with other things they need or they can vendor/fork it and tune to what they need (e.g. with a quick sed) or override via some defines, etc.

This choice can be made for direct use. But it is not so easy for when dependency A wants one "base" and dependency B wants another one. Such clashes are not just hypothetical. vcpkg also modifies the output name of crashpad's vendored (chrome) base lib:
https://github.com/microsoft/vcpkg/blob/acd5bba5aac8b6573b5f6f463dc0341ac0ee6fa4/ports/crashpad/fix-lib-name-conflict-1.patch#L8-L10

> Would adding a patch with something like this work?   
> ... some magic tricks to use custom vcpkg triples that allows e.g. overriding defines?

Triplet files are for script mode. There is a robust interface for triplet files to inject per-port CMake options (`VCPKG_CMAKE_CONFIGURE_OPTIONS`). Injecting flags/definitions into project mode must rely on the toolchain doing the right thing and is less robust.

So if you want to offer configuration of the library name, you could add this to the project:
~~~cmake
set(LIBBASE_OUTPUT_NAME "libbase" CACHE STRING "The library's basename. Modify to resolve name clashes.")
set_target_properties(libbase PROPERTIES OUTPUT_NAME "${LIBBASE_OUTPUT_NAME}")
~~~

Then the portfile could do
~~~cmake
vcpkg_cmake_configure(
   ...
   OPTIONS
      -DLIBBASE_OUTPUT_NAME=ripper37-libbase
)
~~~

RippeR37 committed:
Updated port to change lib output name to avoid conflicts

RippeR37 committed:
Update versions

dg0yt commented on ports/ripper37-libbase/portfile.cmake:
```diff
@@ -0,0 +1,32 @@
+)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(PACKAGE_NAME "ripper37-libbase")
```
~~~
ripper37-libbase:x64-windows-static:/share/ripper37-libbase/libbase-config.cmake
~~~
AFAIU the intented package name is to remain unchanged `libbase`. So the CMake config must be in a directory name starting with `libbase`.
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME "libbase")
```
I didn't check if you need to update the patch and/or add `CONFIG_PATH <where_from>` here.

dg0yt replied:
You port will own two dirs in `share`. It is not so unusual with CMake files.

RippeR37 replied:
Ok, reverted the changes that tried to merge those, package should now be `libbase` as it was originally. Thank you!

I also tested it locally both with `vcpkg install ripper37-libbase` (on modified vcpkg repo) and via custom-overlay in manifest mode to load and use `libbase` in some example project and I found a bug (in original library) that broke building if gtest/benchmark weren't available (I had a wrong impression about using `find_dependency()` calls in cmake config file, I need to fix it...).

For now I added a patch for this that is fine for vcpkg, I'll fix it in the `libbase` in future version.

dg0yt reviewed (commented):
From the CI filelists:
~~~
ripper37-libbase:x64-windows-static:/lib/ripper37-libbase.lib
~~~
:+1:

RippeR37 committed:
Revert package name/dirs in .../share/ to use libbase

RippeR37 committed:
Update versions

RippeR37 committed:
Fix config path

RippeR37 committed:
Update versions

RippeR37 committed:
Fixed a bug which required gtest+benchmark to be installed even if unused

RippeR37 committed:
Update versions

Cheney-W commented:
Usage test passed with x64-windows triplet.

RippeR37 commented:
Huge thank you to @Cheney-W and @dg0yt for all the help here. Sorry it took so long, but I'll do my best to come more prepared for updates and hopefully it will be smoother. Thank you for explaining everything along the way etc. :)
